### PR TITLE
[ADP-2565] Improve performance of `insertMany` with prepared statement

### PR DESCRIPTION
**Pull request by HeinrichApfelmus**
This pull request improves the performance of `insertMany` by using a single prepared statement and setting new parameters multiple times.

### Comments

* I have skipped extensive property testing for now, as I want to stabilize the API first. The module `Demo.Database` provides a basic functionality check.

### Issue Number

 ADP-2565

### CODE REVIEW

File: lib/delta-table/src/Database/Table/SQLite/Simple/Exec.hs [hunk 1]
```diff
@@ -134,8 +134,22 @@ selectWhere expr = queryNamed . Stmt.selectWhere expr
...
+-- | Insert many rows into a database table.
+--
+-- As an optimization, we use a single prepared SQL statement.
 insertMany :: IsTableSql t => [Row t] -> proxy t -> SqlM ()
-insertMany rows proxy = for_ rows (`insertOne` proxy)
+insertMany rows proxy =
+    ReaderT $ \conn -> do
+        -- The 'query' string contains '?' which will be substituted
+        -- for the values in the row.
+        let query = fst $ renderStmt $ Stmt.insertOne proxy
```
Comment (Anviking) [inline lib/delta-table/src/Database/Table/SQLite/Simple/Exec.hs]: If there shouldn't be any bindings, then perhaps
```suggestion
        let (query, _noBindings) = $ renderStmt $ Stmt.insertOne proxy
```
unless we even want it to crash if there are?
```suggestion
        let (query, []) = $ renderStmt $ Stmt.insertOne proxy
```

**Review by Anviking — APPROVED** [review]
👍 Obviously some tests would be nice, but seems to work

File: lib/delta-table/src/Database/Table/SQLite/Simple/Exec.hs [hunk 2]
```diff
@@ -134,8 +134,22 @@ selectWhere expr = queryNamed . Stmt.selectWhere expr
...
 
+-- | Insert many rows into a database table.
+--
+-- As an optimization, we use a single prepared SQL statement.
 insertMany :: IsTableSql t => [Row t] -> proxy t -> SqlM ()
-insertMany rows proxy = for_ rows (`insertOne` proxy)
+insertMany rows proxy =
+    ReaderT $ \conn -> do
+        -- The 'query' string contains '?' which will be substituted
+        -- for the values in the row.
```
Comment (Anviking) [inline lib/delta-table/src/Database/Table/SQLite/Simple/Exec.hs]:
```suggestion
        -- for the values in the row, but no named params.
```

File: lib/delta-table/src/Database/Table/SQLite/Simple/Exec.hs (hunk 1 above)
Comment (HeinrichApfelmus) [reply lib/delta-table/src/Database/Table/SQLite/Simple/Exec.hs]: `_noBindings` sounds good to me.

I don't think that `let (query, []) = …` actually crashes — `let` bindings introduce [irrefutable pattern][1].

  [1]: https://www.haskell.org/tutorial/patterns.html#tut-lazy-patterns

**Pull request merged**